File: checkout-core-card/src/main/java/com/adyen/checkout/core/card/CardType.java
package com.adyen.checkout.core.card;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.adyen.checkout.base.TxVariantProvider;
import com.adyen.checkout.base.internal.JsonObject;
import com.adyen.checkout.base.internal.JsonObject.SerializedName;

import org.json.JSONException;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Copyright (c) 2018 Adyen B.V.
 * <p>
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 * <p>
 * Created by timon on 07/02/2018.
 */
public enum CardType implements TxVariantProvider {
    // NAME("txVariant", "^(first.two)(third)d*$"))
    ACCEL("accel", Pattern.compile("^(19|23|4\\d|5\\d|6[02348]|81)(\\d)\\d*$")),
    AFFN("affn", Pattern.compile("^(4\\d|5[0-8]|6[04])(\\d)\\d*$")),
    @SerializedName("amex")
    AMERICAN_EXPRESS("amex", Pattern.compile("^(3[47])(\\d)\\d*$")),
    ARGENCARD("argencard", Pattern.compile("^(50)(1)\\d*$")),
    BCMC("bcmc", Pattern.compile("^([46]7)([09])\\d*$")),
    @SerializedName("bijcard")
    BIJENKORF_CARD("bijcard", Pattern.compile("^(51)(0)\\d*$")),
    CABAL("cabal", Pattern.compile("^(58|6[03])([03469])\\d*$")),
    CADOCARTE("cadocarte", Pattern.compile("^(48)(9)\\d*$")),
    CARNET("carnet", Pattern.compile("^(28|5[08]|6[023])([246-9])\\d*$")),
    CARTEBANCAIRE("cartebancaire", Pattern.compile("^(18|3[07]|4\\d|5[0-68]|6[2379]|92)(\\d)\\d*$")),
    CENCOSUD("cencosud", Pattern.compile("^(60)(\\d)\\d*$")),
    CHEQUEDEJENEUR("chequedejeneur", Pattern.compile("^(54)(0)\\d*$")),
    CU24("cu24", Pattern.compile("^(4\\d|5[0-58]|6[34])(\\d)\\d*$")),
    CUP("cup", Pattern.compile("^(3[5-7]|4\\d|5[1-5]|6[02389]|9[0145689])(\\d)\\d*$")),
    DANKORT("dankort", Pattern.compile("^(50)([17])\\d*$")),
    DINERS("diners", Pattern.compile("^(36|65)(\\d)\\d*$")),
    DISCOVER("discover", Pattern.compile("^(36|6[045])(\\d)\\d*$")),
    EFTPOS_AUSTRALIA("eftpos_australia", Pattern.compile("^(4\\d|5[0-8]|6[023])(\\d)\\d*$")),
    ELO("elo", Pattern.compile("^(4[035]|50|6[235])([014-9])\\d*$")),
    FORBRUGSFORENINGEN("forbrugsforeningen", Pattern.compile("^(60)(0)\\d*$")),
    GOLDSMITHCARD("goldsmithscard", Pattern.compile("^(98)(2)\\d*$")),
    GOOGLEWALLET("googlewallet", Pattern.compile("^(51|60)([18])\\d*$")),
    HALLMARKCARD("hallmarkcard", Pattern.compile("^(98)(2)\\d*$")),
    HIPER("hiper", Pattern.compile("^(63)(7)\\d*$")),
    HIPERCARD("hipercard", Pattern.compile("^(38|60)([46])\\d*$")),
    INTERLINK("interlink", Pattern.compile("^(21|4\\d|5[0-8]|6[0-4]|92)(\\d)\\d*$")),
    JCB("jcb", Pattern.compile("^(35)([2-8])\\d*$")),
    LASER("laser", Pattern.compile("^(6[37])([07])\\d*$")),
    MAESTRO("maestro", Pattern.compile("^(21|4\\d|5[0-8]|6[0-47]|70|8[24])(\\d)\\d*$")),
    @SerializedName("maestrouk")
    MAESTRO_UK("maestrouk", Pattern.compile("^(6[37])([0135679])\\d*$")),
    MAPPINWEBBCARD("mappinwebbcard", Pattern.compile("^(98)(2)\\d*$")),
    @SerializedName("mc")
    MASTERCARD("mc", Pattern.compile("^(13|2[0-37]|35|4\\d|5[0-8]|6[0-478]|7[01357]|87|9[178])(\\d)\\d*$")),
    MIR("mir", Pattern.compile("^(22)(0)\\d*$")),
    NARANJA("naranja", Pattern.compile("^(37|40|5[28])([279])\\d*$")),
    NETPLUS("netplus", Pattern.compile("^(5[24])([18])\\d*$")),
    NETS("nets", Pattern.compile("^(11|88|99)([189])\\d*$")),
    NYCE("nyce", Pattern.compile("^(19|2[27]|4\\d|5[0-8]|6[02-5])(\\d)\\d*$")),
    PULSE("pulse", Pattern.compile("^(1[01]|22|3[016]|4\\d|5\\d|6[0-589]|81|9[0145689])(\\d)\\d*$")),
    SHAZAM("shazam", Pattern.compile("^(13|4\\d|5[0-8]|6[024])(\\d)\\d*$")),
    SHOPPING("shopping", Pattern.compile("^(27|58|60)([39])\\d*$")),
    SOLO("solo", Pattern.compile("^(67)(6)\\d*$")),
    STAR("star", Pattern.compile("^(4\\d|5[0-8]|6[0-5]|70)(\\d)\\d*$")),
    TROY("troy", Pattern.compile("^(97)(9)\\d*$")),
    UATP("uatp", Pattern.compile("^(1\\d)([0-8])\\d*$")),
    VISA("visa", Pattern.compile("^(4\\d|5[0678]|6[024]|7[034]|97)(\\d)\\d*$")),
    VISADANKORT("visadankort", Pattern.compile("^(45)(7)\\d*$")),
    VISASARAIVACARD("visasaraivacard", Pattern.compile("^(44)(4)\\d*$")),
    WOSCARD("woscard", Pattern.compile("^(98)(2)\\d*$"));

    private final String mTxVariant;

    private final Pattern mPattern;

    /**
     * Get the {@link CardType} for a given {@link TxVariantProvider}.
     *
     * @param txVariantProvider A {@link TxVariantProvider}.
     * @return The {@link CardType}, or {@code null} if no {@link CardType} matches the payment method type.
     */
    @Nullable
    public static CardType forTxVariantProvider(@NonNull TxVariantProvider txVariantProvider) {
        try {
            return JsonObject.parseEnumValue(txVariantProvider.getTxVariant(), CardType.class);
        } catch (JSONException e) {
            return null;
        }
    }

    /**
     * Estimate all potential {@link CardType CardTypes} for a given card number.
     *
     * @param cardNumber The potential card number.
     * @return All matching {@link CardType CardTypes} if the number was valid, otherwise an empty {@link List}.
     */
    @NonNull
    public static List<CardType> estimate(@NonNull String cardNumber) {
        List<CardType> result = new ArrayList<>();

        for (CardType type : CardType.values()) {
            if (type.isEstimateFor(cardNumber)) {
                result.add(type);
            }
        }

        return result;
    }

    CardType(@NonNull String txVariant, @NonNull Pattern pattern) {
        mTxVariant = txVariant;
        mPattern = pattern;
    }

    @NonNull
    @Override
    public String getTxVariant() {
        return mTxVariant;
    }

    /**
     * Returns whether a given card number is estimated for this {@link CardType}.
     *
     * @param cardNumber The card number to make an estimation for.
     * @return Whether the {@link CardType} is an estimation for a given card number.
     */
    public boolean isEstimateFor(@NonNull String cardNumber) {
        String normalizedCardNumber = cardNumber.replaceAll("\\s", "");
        Matcher matcher = mPattern.matcher(normalizedCardNumber);

        return (matcher.matches() || matcher.hitEnd()) && normalizedCardNumber.length() <= CardValidator.NUMBER_MAXIMUM_LENGTH;
    }
}

File: checkout-core/src/main/java/com/adyen/checkout/core/CardType.kt
/*
 * Copyright (c) 2024 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by ozgur on 2/10/2024.
 */

package com.adyen.checkout.core

import java.util.regex.Pattern

/**
 * Predefined list of card schemes. Can be used to configure the supported card schemes with
 * [CardConfiguration.Builder.setSupportedCardTypes].
 */
enum class CardType(val txVariant: String, private val pattern: Pattern) {

    AMERICAN_EXPRESS("amex", Pattern.compile("^3[47][0-9]{0,13}$")),
    ARGENCARD("argencard", Pattern.compile("^(50)(1)\\d*$")),
    BCMC("bcmc", Pattern.compile("^((6703)[0-9]{0,15}|(479658|606005)[0-9]{0,13})$")),
    BIJENKORF_CARD("bijcard", Pattern.compile("^(5100081)[0-9]{0,9}$")),
    CABAL("cabal", Pattern.compile("^(58|6[03])([03469])\\d*$")),
    CARTEBANCAIRE("cartebancaire", Pattern.compile("^[4-6][0-9]{0,15}$")),
    CODENSA("codensa", Pattern.compile("^(590712)[0-9]{0,10}$")),
    CUP("cup", Pattern.compile("^(62|81)[0-9]{0,17}$")),
    DANKORT("dankort", Pattern.compile("^(5019)[0-9]{0,12}$")),
    DINERS("diners", Pattern.compile("^(36)[0-9]{0,12}$")),
    DISCOVER(
        txVariant = "discover",
        pattern = Pattern.compile("^(6011[0-9]{0,12}|(644|645|646|647|648|649)[0-9]{0,13}|65[0-9]{0,14})$")
    ),
    ELO(
        txVariant = "elo",
        pattern = Pattern.compile(
            "^((((506699)|(506770)|(506771)|(506772)|(506773)|(506774)|(506775)|(506776)|(506777)|(506778)|" +
                "(401178)|(438935)|(451416)|(457631)|(457632)|(504175)|(627780)|(636368)|(636297))[0-9]{0,10})|" +
                "((50676)|(50675)|(50674)|(50673)|(50672)|(50671)|(50670))[0-9]{0," + "11})$"
        )
    ),
    FORBRUGSFORENINGEN("forbrugsforeningen", Pattern.compile("^(60)(0)\\d*$")),
    VISAALPHABANKBONUS("visaalphabankbonus", Pattern.compile("^(450903)[0-9]{0,10}$")),
    MCALPHABANKBONUS("mcalphabankbonus", Pattern.compile("^(510099)[0-9]{0,10}$")),
    HIPER("hiper", Pattern.compile("^(637095|637599|637609|637612)[0-9]{0,10}$")),
    HIPERCARD("hipercard", Pattern.compile("^(606282)[0-9]{0,10}$")),
    JCB("jcb", Pattern.compile("^(352[8,9]{1}[0-9]{0,15}|35[4-8]{1}[0-9]{0,16})$")),
    OASIS("oasis", Pattern.compile("^(982616)[0-9]{0,10}$")),
    KARENMILLER("karenmillen", Pattern.compile("^(98261465)[0-9]{0,8}$")),
    WAREHOUSE("warehouse", Pattern.compile("^(982633)[0-9]{0,10}$")),
    LASER("laser", Pattern.compile("^(6304|6706|6709|6771)[0-9]{0,15}$")),
    MAESTRO("maestro", Pattern.compile("^(5[0|6-8][0-9]{0,17}|6[0-9]{0,18})$")),
    MAESTRO_UK("maestrouk", Pattern.compile("^(6759)[0-9]{0,15}$")),
    MASTERCARD("mc", Pattern.compile("^(5[1-5][0-9]{0,14}|2[2-7][0-9]{0,14})$")),
    MIR("mir", Pattern.compile("^(220)[0-9]{0,16}$")),
    NARANJA("naranja", Pattern.compile("^(37|40|5[28])([279])\\d*$")),
    SHOPPING("shopping", Pattern.compile("^(27|58|60)([39])\\d*$")),
    SOLO("solo", Pattern.compile("^(6767)[0-9]{0,15}$")),
    TROY("troy", Pattern.compile("^(97)(9)\\d*$")),
    UATP("uatp", Pattern.compile("^1[0-9]{0,14}$")),
    VISA("visa", Pattern.compile("^4[0-9]{0,18}$")),
    VISADANKORT("visadankort", Pattern.compile("^(4571)[0-9]{0,12}$"));

    /**
     * Returns whether a given card number is estimated for this [CardType].
     *
     * @param cardNumber The card number to make an estimation for.
     * @return Whether the [CardType] is an estimation for a given card number.
     */
    internal fun isEstimateFor(cardNumber: String): Boolean {
        val normalizedCardNumber = cardNumber.replace("\\s".toRegex(), "")
        val matcher = pattern.matcher(normalizedCardNumber)
        return matcher.matches() || matcher.hitEnd()
    }

    companion object {

        /**
         * Get [CardType] from the brand name as it appears in the Checkout API.
         */
        fun getByBrandName(brand: String): CardType? {
            return values().firstOrNull { it.txVariant == brand }
        }
    }
}
